src/Symfony/Contracts/HttpClient/HttpClientInterface.php

```diff
@@ -0,0 +1,88 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\Test\HttpClientTest;
+
+/**
+ * Provides flexible methods for requesting HTTP resources.
+ *
+ * @see HttpClientTest for a reference test suite
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @experimental in 1.1
+ */
+interface HttpClientInterface
+{
+    public const OPTIONS_DEFAULTS = [
+        'auth' => null,         // string - a username:password enabling HTTP Basic authentication
+        'query' => [],          // array - associative array of query string values to merge with the request's URL
+        'headers' => [],        // iterable|string[]|string[][] - headers names provided as keys or as part of values
+        'body' => '',           // array|string|resource|\Traversable|\Closure - the callback SHOULD yield a string
+                                //   smaller than the amount requested as argument; the empty string signals EOF; when
+                                //   an array is passed, it is meant as a form payload of field names and values
+        'json' => null,         // array|\JsonSerializable - when set, implementations MUST set the "body" option to
+                                //   the JSON-encoded value and set the "content-type" headers to a JSON-compatible
+                                //   value it is they are not defined - typically "application/json"
+        'user_data' => null,    // mixed - any extra data to attach to the request (scalar, callable, object...) that
+                                //         MUST be available via $response->getInfo('data') - not used internally
+        'max_redirects' => 20,  // int - the maximum number of redirects to follow; a value lower or equal to 0 means
+                                //       redirects should not be followed; "Authorization" and "Cookie" headers MUST
+                                //       NOT follow except for the initial host name
+        'http_version' => null, // string - defaults to the best supported version, typically 1.1 or 2.0
+        'base_uri' => null,     // string - the URI to resolve relative URLs, following rules in RFC 3986, section 2
+        'buffer' => true,       // bool - whether the content of the response should be buffered or not
+        'on_progress' => null,  // callable - called on download/upload at least once per second with arguments
+                                //            int $dlNow, int $dlSize and array $info; any exceptions thrown from the
+                                //            callback MUST abort the request
+        'resolve' => [],        // string[] - a map of host to IP address that SHOULD replace DNS resolution
```

Comment:
would be nice to allow a callback here which take a domain name as first parameter and returns an array of ip, so someone can write its own dns resolver. But can me done in another PR

Reply:
Nice idea! Right now, you can provide only one IP per host, dunno if we need to support more, even when provided by the callable. I'll let you send the PR after merge if you don't  mind :) (or create an issue if you prefer)

Reply:
Oh yes, believe it was an array of ip, so callback can return only one ip (even if there is multiple it should be its reponsaiblity to return only one)